### PR TITLE
Fix build issues and add two libraries

### PR DESCRIPTION
**Pull request by beeselmane**
The first few commits are related to compilation errors/warnings I was having on my laptop because of my case insensitive filesystem and OS version mismatch.

The final commits do three main things:
1. Install as a 'DeveloperInternal' component the private headers and libraries built with the system. This is how Apple builds Xcode SDKs and is likely to be best practice when building a standalone OS. This allows users to build against internal APIs on a given version, but can also be easily disabled by deleting the directory tree at `/usr/local`, as well as removing `*.framework/PrivateHeaders` where it exists. This change affects XNU headers, installing the remaining built frameworks, libsystem_kernel, and AvailabilityVersions.
2. I've cleaned up and tagged libsystem_kernel where there were inconsistencies between the Xcode project and the CMake scripts as they existed before. The library still builds in both the static and dynamic variants, but some source files have been changed to the correct variant, and some header files which were not meant to be globally installed are no longer gloablly installed. Also, in line with (1), private headers are properly installed to the directory tree under `/usr/local`. Building from a clean build directory is working on my end with no issues.
3. Added `libCrashReporterClient.a` The implementation I added was taken from the PureDarwin/libcoreservices repo, and builds fine. I've added the dependency in `libsystem_kernel.dylib`, as Apple seems to do on their systems.

Any notes or changes are more than welcome!

### CODE REVIEW

**Comment by beeselmane** [comment]
I've found some missing source files in libsystem_kernel, along with some MIG interfaces not being generated by CMake. I'll update this pull request if I can get them to build.

**Comment by wjk** [comment]
❤️ ❤️ ❤️  

I would appreciate it if you would rebase onto latest main; GitHub says there are conflicts. I can then start the automated CI and build it on my end. Regarding the private headers, your installation of them may now be against our CMake coding standards. If you still believe that it is appropriate to publish them, please let me know why. Thanks!

**Comment by wjk** [comment]
> I've found some missing source files in libsystem_kernel, along with some MIG interfaces not being generated by CMake. I'll update this pull request if I can get them to build.

I skipped these files because either their presence broke the build, or if Apple’s build system didn’t build them either. Please elaborate.

**Comment by beeselmane** [comment]
With regard to the first comment, I'll do the rebase, I think it might have been one of the recent commits you had just pushed. For the private headers, I did put them in a separate component, namely `DeveloperInternal`. I think it makes sense to at the very least have the option to include them in a full system build, as which they expose internal interfaces, they could be useful for developers wanting to play with system internals or for running various tests on the resulting system (dyld for instance can be configured to install various tools to modify the shared cache into `/usr/local/bin`, and there are some scripts and things which can be useful when installed into `/usr/local/libexec`. I agree that it makes sense to keep these things out of the default build, I figure by default we can simply have CMake build either just `BaseSystem` or `BaseSystem DeveloperTools`, which currently will exclude private headers for all of the various things we build now.

As for the second comment, I've almost finished getting everything working with regard to the missing files, down to some duplicate symbol issues I believe I have a fix for. Basically, the remaining files rely on the internal MIG headers that we don't currently build. I'll have everything in my next few commits.

**Comment by wjk** [comment]
You raise a very good point regarding the internal headers. I completely forgot about that. I will update the coding style document immediately.

**Comment by wjk** [comment]
And I just fixed the merge conflict for you, if that’s OK. 😅

**Comment by beeselmane** [comment]
Oh thanks! Sounds good, lemme see if I can fixup these MIG files real quick too

**Comment by beeselmane** [comment]
@wjk This should work now. Feel free to try it out on your end. I looked at how the MIG files are supposed to be built from the xcodeproject and now I've included all of the non-stub source files in the final build.

**Comment by beeselmane** [comment]
Also, as I note, I migrated the include/ and private_include/ directories we use to be full of symlinks instead of direct copies, my thinking on this was that if we need to update the projects it makes it much easier to just drop in the new code without having to copy the right headers out and everything. Includes are still centralized, but the underlying headers are referenced from the codebase itself.

File: src/Libraries/libSystem/libsystem_kernel/CMakeLists.txt [hunk 1]
```diff
@@ -234,15 +400,64 @@ target_sources(libsystem_kernel PRIVATE
...
+target_include_directories(libsystem_kernel PRIVATE libsystem_kernel_private_headers)
+
+target_compile_options(libsystem_kernel PRIVATE -fdollars-in-identifiers -no-cpp-precomp -fno-common -fno-stack-protector -fno-stack-check -momit-leaf-frame-pointer -Wno-nullability-completeness -Wno-implicit-function-declaration)
+target_compile_definitions(libsystem_kernel PRIVATE _FORTIFY_SOURCE=0 __DARWIN_VERS_1050=1 NO_SYSCALL_LEGACY SYSCALL_PRE1050)
+
 target_link_libraries(libsystem_kernel PRIVATE libsystem_kernel_static)
-target_link_options(libsystem_kernel PRIVATE -umbrella System -all_load -nostdlib)
-install(TARGETS libsystem_kernel DESTINATION /usr/lib/system COMPONENT BaseSystem)
+target_link_libraries(libsystem_kernel PRIVATE CrashReporterClient)
+target_link_options(libsystem_kernel PRIVATE -umbrella System -all_load)
```
Comment (wjk) [inline src/Libraries/libSystem/libsystem_kernel/CMakeLists.txt:419]: `-nostdlib` needs to be put back.

File: src/Libraries/libSystem/libsystem_kernel/Libsyscall.xcconfig [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#include "<DEVELOPER_DIR>/Makefiles/CoreOS/Xcode/BSD.xcconfig"
```
Comment (wjk) [inline src/Libraries/libSystem/libsystem_kernel/Libsyscall.xcconfig:1]: Could you please delete this file and the Xcode project as well? (You can use `git rm --cached` to keep the file on disk, but remove it from the PR.)

**Review by wjk — DISMISSED** [review]
Just a few last comments. This PR is amazing! 🤩

**Comment by beeselmane** [comment]
With regard to -nostdinc -nostdlib, these flags should be default on all of the stuff we build for the destination system. -nostdlib prevents us from linking target system libraries with host system libraries, and -nostdinc prevents us from compiling against host system headers. They should as such probably be the default for target system code, and we just have to make sure to build everything in the right order with the dependencies properly resolved. I will add the Libc sources to remove the need for `bootstrap_include`, and then I'll try and get the static libsystem libraries built for dyld, and attempt to build libdyld in a separate PR/at a later date. For now though, it'd probably be best to leave the -nostdlib -nostdinc flags on compiling/linking by default for the target.

**Comment by beeselmane** [comment]
I've pulled in the libraries we need for system headers that get libsystem_kernel to build. There are a few dependencies on headers provided by clang. Integrating LLVM is certainly one of the next steps here, for both the clang headers and libcompiler_rt, which will help with the rest of libSystem.

**Comment by wjk** [comment]
> we just have to make sure to build everything in the right order with the dependencies properly resolved

That was exactly how I was doing it before, but there was too much consternation amongst the community on whether or not people could contribute due to this policy, because people wanted to work on parts that did not have their dependencies fully resolved at the time. I am going to make a quick edit to make nostdinc opt-out, and then we can move on as before.

**Comment by wjk** [comment]
Please delete `src/Libraries/architecture/Makefile`. (I would have commented on the file itself, but the diff has become too big for GitHub to handle.

File: src/Libraries/libSystem/libsystem_kernel/CMakeLists.txt (hunk 1 above)
Comment (wjk) [reply src/Libraries/libSystem/libsystem_kernel/CMakeLists.txt:426]: Closing resolved stale conversation. (GitHub doesn’t call it as such, but that’s what it is.)

**Comment by beeselmane** [comment]
Alright, seems to be all ready to me, unless you have anything else for this PR. Further changes would likely be out of the scope for this one.

**Comment by beeselmane** [comment]
> > we just have to make sure to build everything in the right order with the dependencies properly resolved
> 
> That was exactly how I was doing it before, but there was too much consternation amongst the community on whether or not people could contribute due to this policy, because people wanted to work on parts that did not have their dependencies fully resolved at the time. I am going to make a quick edit to make nostdinc opt-out, and then we can move on as before.
> 
> EDIT: This has been done, and the coding style document has been updated to reflect the changes.

As an aside, this does make sense to me, at least until we can get some more of the core system done. I agree with the default being to not include host headers/libraries, though.

**Review by wjk — APPROVED** [review]

**Pull request merged**